components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +72,154 @@ impl MutationObserver {
         Ok(observer)
     }
+
+    /// https://dom.spec.whatwg.org/#queue-a-mutation-observer-compound-microtask
+    /// Queue a Mutation Observer compound Microtask.
+    pub fn queueMutationObserverCompoundMicrotask(compoundMicrotask: Microtask) {
+        // Step 1
+        if ScriptThread::is_mutation_observer_compound_microtask_queued() {
+            return;
+        }
+        // Step 2
+        ScriptThread::set_mutation_observer_compound_microtask_queued(true);
+        // Step 3
+        ScriptThread::enqueue_microtask(compoundMicrotask);
+    }
+
+    /// https://dom.spec.whatwg.org/#notify-mutation-observers
+    /// Notify Mutation Observers.
+    pub fn notifyMutationObservers() {
+        // Step 1
+        ScriptThread::set_mutation_observer_compound_microtask_queued(false);
+        // Step 2
+        let notifyList = ScriptThread::get_mutation_observer();
+        // Step 3, Step 4 not needed as Servo doesn't implement anything related to slots yet.
+        // Step 5: Ignore the specific text about execute a compound microtask.
+        // Step 6 not needed as Servo doesn't implement anything related to slots yet.
+    }
+
+    //https://dom.spec.whatwg.org/#queueing-a-mutation-record
+    //Queuing a mutation record
+    pub fn queue_a_mutation_record(&self, target: &Node, attr_type: Mutation) {
+        let mut given_name = "";
+        let mut given_namespace = "";
+        let mut old_value = "";
+        // Step 1
+        let mut interestedObservers: Vec<JS<MutationObserver>> = vec![];
+        let mut pairedStrings: Vec<DOMString> = vec![];
+        // Step 2
+        let mut nodes: Vec<Root<Node>> = vec![];
+        for ancestor in target.inclusive_ancestors() {
+            nodes.push(ancestor);
+        }
+        // Step 3
+        for node in &nodes {
+            for registered_observer in node.registered_mutation_observers().borrow().iter() {
+                match attr_type {
+                    Mutation::Attribute { ref name, ref namespace, ref oldValue, ref newValue } => {
+                        let given_name = &*name.clone();
+                        let given_namespace = &*namespace.clone();
+                        let old_value = &*oldValue.clone();
+                    },
+                }
+                let condition1: bool = node != &Root::from_ref(target) &&
+                    !registered_observer.subtree.get();
+                let condition2: bool = given_name == "attribute" &&
+                    registered_observer.attributes.get() == false;
+                let condition3: bool = (given_name == "attribute" &&
+                    registered_observer.attribute_filter != DOMRefCell::new(vec![])) &&
+                    (given_namespace != "");
+                let condition4: bool = given_name == "characterData" &&
+                    registered_observer.character_data.get() == false;
+                let condition5: bool = given_name == "childList" &&
```

Comment:
Just to be clear about my earlier comments - `given_name` is entirely unrelated to the specification variable "type", so comparing it against values like "childList" or "attribute" does not make sense. `given_name` is the name of the attribute that was mutated; that is all.